kernel/src/snapshot.rs

```diff
@@ -43,26 +43,33 @@ impl LogSegment {
     /// `read_schema` is the schema to read the log files with. This can be used
     /// to project the log files to a subset of the columns.
     ///
-    /// `predicate` is an optional expression to filter the log files with.
+    /// `meta_predicate` is an optional expression to filter the log files with. It is _NOT_ the
+    /// query's predicate, but rather a predicate for filtering log files themselves.
     #[cfg_attr(feature = "developer-visibility", visibility::make(pub))]
     #[cfg_attr(not(feature = "developer-visibility"), visibility::make(pub(crate)))]
     fn replay(
         &self,
         engine: &dyn Engine,
         commit_read_schema: SchemaRef,
         checkpoint_read_schema: SchemaRef,
-        predicate: Option<Expression>,
+        meta_predicate: Option<Expression>,
     ) -> DeltaResult<impl Iterator<Item = DeltaResult<(Box<dyn EngineData>, bool)>> + Send> {
         let json_client = engine.get_json_handler();
-        // TODO change predicate to: predicate AND add.path not null and remove.path not null
         let commit_stream = json_client
-            .read_json_files(&self.commit_files, commit_read_schema, predicate.clone())?
+            .read_json_files(
+                &self.commit_files,
+                commit_read_schema,
+                meta_predicate.clone(),
+            )?
             .map_ok(|batch| (batch, true));
 
         let parquet_client = engine.get_parquet_handler();
-        // TODO change predicate to: predicate AND add.path not null
```

Comment:
I removed this TODO because the P&M query also invokes this code path, and filtering by `add.path NOT NULL` is just plain incorrect. Also, see the code comment at the other `replay` call site for why it doesn't make sense to pass `add.path IS NOT NULL` as a row group skipping filter.